pkg/admission/zone_tracker_test.go

```diff
@@ -286,3 +294,125 @@ func TestLastDownscaledNonExistentZone(t *testing.T) {
 		t.Errorf("lastDownscaled did not handle non-existent zone correctly")
 	}
 }
+
+func TestZoneTrackerConcurrentDownscale(t *testing.T) {
+	f := newFakeHttpClient(func(r *http.Request) (*http.Response, error) {
+		return &http.Response{
+			StatusCode: http.StatusOK,
+			Body:       io.NopCloser(bytes.NewBuffer([]byte(""))),
+		}, nil
+	})
+
+	logger := newDebugLogger()
+
+	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		w.WriteHeader(http.StatusOK)
+	}))
+	defer ts.Close()
+
+	u, err := url.Parse(ts.URL)
+	require.NoError(t, err)
+	require.NotEmpty(t, u.Port())
+
+	path := "/prepare-downscale"
+	rolloutGroupIngester := "ingester"
+	ingesterZoneA := "ingester-zone-a"
+	rolloutGroupIndexGateway := "index-gateway"
+	indexGatewayZoneA := "index-gateway-zone-a"
+
+	namespace := "test"
+	dryRun := false
+	buildAdmissionRequest := func(rolloutGroup string, stsName string) admissionv1.AdmissionReview {
+		oldParams := templateParams{
+			Replicas:          5,
+			DownScalePathKey:  config.PrepareDownscalePathAnnotationKey,
+			DownScalePath:     path,
+			DownScalePortKey:  config.PrepareDownscalePortAnnotationKey,
+			DownScalePort:     u.Port(),
+			DownScaleLabelKey: config.PrepareDownscaleLabelKey,
+			RolloutGroup:      rolloutGroup,
+		}
+
+		newParams := templateParams{
+			Replicas:          2,
+			DownScalePathKey:  config.PrepareDownscalePathAnnotationKey,
+			DownScalePath:     path,
+			DownScalePortKey:  config.PrepareDownscalePortAnnotationKey,
+			DownScalePort:     u.Port(),
+			DownScaleLabelKey: config.PrepareDownscaleLabelKey,
+			RolloutGroup:      rolloutGroup,
+		}
+
+		rawObject, err := statefulSetTemplate(newParams)
+		require.NoError(t, err)
+
+		oldRawObject, err := statefulSetTemplate(oldParams)
+		require.NoError(t, err)
+
+		return admissionv1.AdmissionReview{
+			Request: &admissionv1.AdmissionRequest{
+				Kind: metav1.GroupVersionKind{
+					Group:   "apps",
+					Version: "v1",
+					Kind:    "Statefulset",
+				},
+				Resource: metav1.GroupVersionResource{
+					Group:    "apps",
+					Version:  "v1",
+					Resource: "statefulsets",
+				},
+				Name:      stsName,
+				Namespace: namespace,
+				Object: runtime.RawExtension{
+					Raw: rawObject,
+				},
+				OldObject: runtime.RawExtension{
+					Raw: oldRawObject,
+				},
+				DryRun: &dryRun,
+			},
+		}
+	}
+
+	api := fake.NewSimpleClientset()
+
+	zt := newZoneTracker(api, namespace, "testconfigmap")
+
+	// mimic in progress downscaling request of rollout group ingester
+	zt.rolloutGroupDownscalingInProgress.Store(rolloutGroupIngester, ingesterZoneA)
```

Comment:
Rather than reaching into the internals like this, what if we changed the test to something like:

1. configure StatefulSet to call fake prepare downscale handler
2. initiate downscale, rollout-operator calls fake handler and marks downscale as in progress
3. fake handler does not return until signalled later
4. initiate a second downscale: rollout-operator should reject the call immediately given the other is in progress
5. release fake handler to allow first request to complete